index.d.ts

```diff
@@ -0,0 +1,49 @@
+declare module 'nsfw' {
+    /** Returns a Promise that resolves to the created NSFW Object.
+     * @param {watchPath} watchPath - the path that nsfw should watchPath
+     * @param {eventCallback} eventCallback - callback that will be fired when NSFW has change events
+     * @param {options} options - options
+     */
+    interface NsfwFunction {
+        (watchPath: string, eventCallback: (events: Array<nsfw.ChangeEvent>) => void, options: nsfw.Options): Promise<nsfw.NSFW>;
+        actions: {
+            CREATED: number;
+			DELETED: number;
+			MODIFIED: number;
+			RENAMED: number;
+        };
+    }
+    
+    namespace nsfw {
+        interface Options {
+            /** time in milliseconds to debounce the event callback */
+            debounceMS?: number;
+            /**  callback to fire in the case of errors */
+            errorCallback: (err: any) => void
+        }
+
+        interface ChangeEvent {

```

Comment:
I feel like this could exhibit much stronger typing behavior if we tried implementing the interface like this:
```typescript
enum ActionType {
    CREATED,
    DELETED,
    MODIFIED,
    RENAMED
}

interface RenamedFileEvent {
    action: ActionType.RENAMED;
    directory: string;
    oldFile: string;
    newFile: string;
    newDirectory: string;
}

interface GenericFileEvent<T extends ActionType> {
    action: T;
    directory: string;
    file: string;
}

type CreatedFileEvent = GenericFileEvent<ActionType.CREATED>;
type DeletedFileEvent = GenericFileEvent<ActionType.DELETED>;
type ModifiedFileEvent = GenericFileEvent<ActionType.MODIFIED>;

type FileChangeEvent = CreatedFileEvent | DeletedFileEvent | ModifiedFileEvent | RenamedFileEvent;
```

Reply:
If we use the proposed typing behavior, you should cast the event to the specific one. Here is an example:
![image](https://user-images.githubusercontent.com/5815794/64673775-c3a3cf00-d477-11e9-938f-8c56273acfae.png)
![image](https://user-images.githubusercontent.com/5815794/64673559-2a74b880-d477-11e9-8b2f-bf9b3bb6b83c.png)
![image](https://user-images.githubusercontent.com/5815794/64673736-a0791f80-d477-11e9-833a-129a32eadbbb.png)

Reply:
You actually don't need to cast it at all. Typescript is more than capable of inferring the types by leveraging the action type. For instance, the following code is sound, and typescript is able to recognize that only a FileChangedEvent with ActionType.RENAMED should have newFile on it. I would definitely argue against using less precise mechanics such as _every_ event has an optional oldFile event, when in fact that is not the case. The typing should reflect the actual datastructure and be in agreement with the documentation as much as possible.

See:
http://www.typescriptlang.org/play/#code/KYOwrgtgBAggxgFwJYHsQBUCeAHYUDeAUFCVAMIBKAojOlQCIA0xp9VAMlXUyyQLIB5egEkAYsIbNSUagDkYfBoQC+hQkhAJgAJwBmAQzh4KofRGAATUUgA2wKgDdQCAryiHkaAFyxEqDDjAAHRyCgwA3G4WSNrAiCjamD4AzgjaGgDmkdIoNla2wClpmdmkIMAA7tZ2RekgWW7lFfQxcQgJSVCpdQ2q6po6BkZQAOKgOkhw1fZOmgA86FDAAB5aIBbJvp4BuAB8rtIe-j7opSTRsfGJtSVuugU39ZF9CIHksfpa+XaOzlAAvKNxukpgVfvN4NssLggpQaNxdpFXrgoPRgHYvtNwS5AWNyiCsbMEHNIf5ocE2JwEUi3nwUNF7pZCX9ccDJsyIX40OSgoIROIGIi1Mi8NMyAALfT1GYs97AT5MsFEqAAH1R6OAmKVfzVdIZSEVP2VapMIDMhplmkihDgaFSUC09sBAApgETkj4xZLpdiANoAXQAlI8MgGAfsiNJbSB7fc7LJzR6usV6mHAQGzlBdAkoM6MVAkACoAAGcIFqBzJbuoJ2eoIcVlgDUjaQgYO0lI0ftbtlPc0yV9SH9mekyQqSAQcHFub7CCCRzQbcjHZXcH0yTwpO5gRCVHkinoXjcK5XceACfMySC2DAyXFrqJQSa00DI5P0gARh8ANZv1frzcuR2YI4VoBgj3fKMAK2Mkd0pLhwOPE81w3GDtxhPkxAkQ8kMgrMCgvYArxvO8H2cIIz1fXDIK-eVf2o1RpEY0hYgQMBtBAfD40TZ4gA

Reply:
That's pretty neat, isn't it!

Reply:
Yes, I agree with you. I've committed the latest changes based on your comments.